File: iosApp/iosApp/Manager/RecipeRepositoryImpl.swift
import CoreData
import Shared
import Foundation

class RecipeRepositoryImpl: RecipeRepository {
    private let context: NSManagedObjectContext
    private let apiService: ApiService

    init(context: NSManagedObjectContext, apiService: ApiService) {
        self.context = context
        self.apiService = apiService
    }

    // Synchronize data from API to Core Data
    func syncData() async -> Bool {
        do {
            // Fetch categories
            let categories = try await apiService.fetchRecipeCategories()
            try saveCategories(categories: categories)

            // Fetch measurement units
            let units = try await apiService.fetchMeasurementUnits()
            try saveUnits(units: units)

            return true
        } catch {
            print("Failed to sync data: \(error)")
            return false
        }
    }

    private func saveCategories(categories: [RecipeCategory]) throws {
        let categoryEntities = categories.map { category -> RecipeCategoryEntity in
            let entity = RecipeCategoryEntity(context: self.context)
            entity.id = Int32(category.id)
            entity.displayName = category.displayName
            return entity
        }

        let existingCategories = try context.fetch(RecipeCategoryEntity.fetchRequest())
        existingCategories.forEach { context.delete($0) }

        categoryEntities.forEach { context.insert($0) }
        try context.save()
    }

    private func saveUnits(units: [MeasurementUnit]) throws {
        let unitEntities = units.map { unit -> MeasurementUnitEntity in
            let entity = MeasurementUnitEntity(context: self.context)
            entity.id = Int32(unit.id ?? 0)
            entity.displayName = unit.displayName
            entity.symbol = unit.symbol
            return entity
        }

        let existingUnits = try context.fetch(MeasurementUnitEntity.fetchRequest())
        existingUnits.forEach { context.delete($0) }

        unitEntities.forEach { context.insert($0) }
        try context.save()
    }

    func saveRecipe(recipe: Recipe) async -> Bool {
        guard recipe.isPrivate else {
            return false
        }

        let recipeEntity = RecipeEntity(context: self.context)
        recipeEntity.id = Int32(recipe.id)
        recipeEntity.userID = recipe.userID
        recipeEntity.title = recipe.title
        recipeEntity.recipeCategoryId = Int32(recipe.recipeCategory.id)
        recipeEntity.portionSize = recipe.portionSize
        recipeEntity.preparationTime = recipe.preparationTime
        recipeEntity.cookingTime = recipe.cookingTime
        recipeEntity.preparation = recipe.preparation
        recipeEntity.rating = Int32(recipe.rating)
        recipeEntity.notes = recipe.notes
        recipeEntity.titleImage = recipe.titleImage
        recipeEntity.isPrivate = recipe.isPrivate
        recipeEntity.created = recipe.created
        recipeEntity.lastUpdated = recipe.lastUpdated

        recipe.ingredients.forEach { ingredient in
            let ingredientEntity = IngredientEntity(context: self.context)
            ingredientEntity.id = Int32(ingredient.id)
            ingredientEntity.value = ingredient.value
            ingredientEntity.name = ingredient.name
            ingredientEntity.formId = ingredient.form?.id.flatMap { Int32($0) }
            ingredientEntity.unitId = Int32(ingredient.unit.id)
            context.insert(ingredientEntity)
        }

        do {
            try context.save()
            return true
        } catch {
            print("Failed to save recipe: \(error)")
            return false
        }
    }
}

File: shared/src/androidMain/kotlin/RecipeRepositoryImpl.kt

import database.IngredientDao
import database.IngredientEntity
import database.MeasurementUnitDao
import database.MeasurementUnitEntity
import database.RecipeCategoryDao
import database.RecipeCategoryEntity
import database.RecipeDao
import database.RecipeEntity
import model.recipe.Recipe
import network.ApiService

class RecipeRepositoryImpl(
    private val apiService: ApiService,
    private val recipeDao: RecipeDao,
    private val recipeCategoryDao: RecipeCategoryDao,
    private val measurementUnitDao: MeasurementUnitDao,
    private val ingredientDao: IngredientDao,
) : RecipeRepository {

    suspend fun syncData() {
        val categories = apiService.fetchRecipeCategories()
        val units = apiService.fetchMeasurementUnits()

        recipeCategoryDao.insertCategories(categories.map { RecipeCategoryEntity(it.id, it.displayName) })

        units.map { measurementUnitDao.insertUnit(MeasurementUnitEntity(it.id!!, it.displayName, it.symbol)) }
    }

    suspend fun saveRecipe(recipe: Recipe) {
        if (recipe.isPrivate) {
            val recipeEntity = RecipeEntity(
                id = recipe.id,
                userID = recipe.userID,
                title = recipe.title,
                recipeCategoryId = recipe.recipeCategory.id,
                portionSize = recipe.portionSize,
                preparationTime = recipe.preparationTime,
                cookingTime = recipe.cookingTime,
                preparation = recipe.preparation,
                rating = recipe.rating,
                notes = recipe.notes,
                titleImage = recipe.titleImage,
                isPrivate = recipe.isPrivate,
                created = recipe.created,
                lastUpdated = recipe.lastUpdated
            )
            recipeDao.insertRecipe(recipeEntity)

            recipe.ingredients.forEach { ingredient ->

                ingredient.unit.id?.let {
                    val ingredientEntity = IngredientEntity(
                        id = ingredient.id,
                        value = ingredient.value,
                        name = ingredient.name,
                        formId = ingredient.form?.id,
                        unitId = it
                    )
                    ingredientDao.insertIngredient(ingredientEntity)
                }

            }
        }
    }

    override suspend fun getRecipes(): List<Recipe> {
        TODO("Not yet implemented")

    }

    override suspend fun getRecipeById(id: Int): Recipe? {
        TODO("Not yet implemented")

    }

    override suspend fun addRecipe(recipe: Recipe): Recipe? {
        TODO("Not yet implemented")
    }

    override suspend fun editRecipe(recipe: Recipe): Recipe? {
        TODO("Not yet implemented")
    }

    override suspend fun deleteRecipe(id: Int): Boolean {
        TODO("Not yet implemented")
    }

    override suspend fun syncRecipes() {
        TODO("Not yet implemented")
    }
}


